samza-core/src/main/java/org/apache/samza/clustermanager/ClusterResourceManager.java

```diff
@@ -105,15 +105,21 @@ public ClusterResourceManager(Callback callback) {
   public abstract void releaseResources(SamzaResource resource);
 
   /***
-   * Requests the launch of a StreamProcessor with the specified context on the resource.
+   * Requests the launch of a StreamProcessor with the specified context on the resource asynchronously.
```

Comment:
Thanks for taking a look Jake!
1. An instance of the `ClusterResourceManager` is always created by passing a `callback`. This is why it's an abstract class.

2. Each `SamzaResource` is a unique identifier for a request. A success is indicated by invoking callback.onSuccess(Resource) and callback.onFailure(Resource, exception) right?
Are you saying that it's not sufficient and proposing a per-method callback or a future instead?

3. The current design is heavily mirrors the design of Hadoop's `NMClientAsync` and `AMRMClientAsync` for the NM and RM respectively.

Reply:
Yes, I was suggesting that having the callback as a method parameter is more flexible. That way they can customize how the callbacks are handled. e.g. as a caller of this method, I may want to monitor the success/failure in order to update my own state and then pass the result success/failure on to some delegate.

Reply:
Here's how I view the requirements for this design:
1. Provide a means to start a `StreamProcessor` on a resource.
2. Provide a way to know if request [1] was successful and the container started.
3. Provide a way to know if request [1] was unsuccessful, and if so, what the error was.
Additionally, It'd be nice to integrate with Yarn (and Mesos later) with minimal effort - since, these are the most popular cluster managers and should serve as good validations that our API is reasonable.

>>  I was suggesting that having the callback as a method parameter is more flexible. That way they can customize how the callbacks are handled. e.g. as a caller of this method, I may want to monitor the success/failure in order to update my own state and then pass the result success/failure on to some delegate.

Agreed on the flexibility. Wouldn't this mean that we will have to track the per-method callback state in `ClusterResourceManager` and `ContainerProcessManager` - just to adapt the APIs of the underlying ClusterManager to this model? 

Given the above requirements, and from my initial analysis of Yarn and Mesos's APIs, I was not convinced that this was absolutely necessary - since, both of them don't support it.

Reply:
If you feel strongly about this, I won't stand in the way. I just find it confusing (and expect it will be hard to maintain the code) when the request/response flow is:
Request: AbstractContainerAllocator -> ClusterManager.launchStreamProcessor()
Response: ClusterManager -> ClusterResourceManager.onStreamProcessorLaunch*()

How would an engineer implementing a new ClusterManager (e.g. Mesos) know which object to invoke the callback on?

That would be like you asking me to make a code change and then expecting that I'll tell Yi when I've finished your request.